dask_kubernetes/utils.py

```diff
@@ -56,19 +56,34 @@ async def get_external_address_for_scheduler_service(
         nodes = await core_api.list_node()
         host = nodes.items[0].status.addresses[0].address
     elif service.spec.type == "ClusterIP":
-        try:
-            # Try to resolve the service name. If we are inside the cluster this should succeeed.
-            host = f"{service.metadata.name}.{service.metadata.namespace}"
-            socket.getaddrinfo(host, port)
-        except socket.gaierror:
-            # If we are outside it will fail and we need to port forward the service.
+        # Try to resolve the service name. If we are inside the cluster this should succeed.
+        host = f"{service.metadata.name}.{service.metadata.namespace}"
+        success = _is_service_available(
+            host=host, port=port, retries=service_name_resolution_retries
+        )
+
+        # If we are outside it will fail and we need to port forward the service.
+        if not success:
             host = "localhost"
             port = await port_forward_service(
                 service.metadata.name, service.metadata.namespace, port
             )
     return f"tcp://{host}:{port}"
 
 
+def _is_service_available(host, port, retries=20):
+    success = False
+    while retries > 0:
+        try:
+            socket.getaddrinfo(host, port)
+            success = True
+        except socket.gaierror:
+            time.sleep(0.5)
+            retries -= 1
+
+    return success
```

Comment:
This could be done a little differently to avoid the counting boilerplate. I would also reraise the exception again so that we don't have to change any other logic.

```suggestion
def _is_service_available(host, port, retries=20):
    for i in range(retries):
        try:
            return socket.getaddrinfo(host, port)
        except socket.gaierror as e:
            if i >= retries - 1:
                raise e
            time.sleep(0.5)
```

Reply:
This is great approach. Thanks!